### PR TITLE
Adds type hinting for `VersionOrder` class

### PR DESCRIPTION
**Pull request by schuylermartin45**
### Description

- Adds type hinting discussed in Issue #13380
- Also adds type hints for `InvalidVersionSpec` and related exceptions

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

### CODE REVIEW

**Comment by travishathaway** [comment]
@conda-bot check

**Review by dholth — APPROVED** [review]

File: conda/models/version.py [hunk 1]
```diff
@@ -243,25 +243,29 @@ def __init__(self, vstr):
...
-    def _eq(self, t1, t2):
+    def _eq(self, t1: list[str], t2: list[str]) -> bool:
         for v1, v2 in zip_longest(t1, t2, fillvalue=[]):
             for c1, c2 in zip_longest(v1, v2, fillvalue=self.fillvalue):
                 if c1 != c2:
                     return False
         return True
 
-    def __eq__(self, other):
+    def __eq__(self, other: object) -> bool:
```
Comment (travishathaway) [inline conda/models/version.py:259]: Wouldn't `Any` be better choice here? I'm wondering why you went with object.

Comment (dholth) [reply conda/models/version.py:259]: LGTM as the input and not the return type: https://stackoverflow.com/questions/39817081/typing-any-vs-object

Comment (kenodegard) [reply conda/models/version.py:259]: hm interesting, well I'm all for replacing `Any` with `object` throughout conda, one less import

**Comment by schuylermartin45** [comment]
Current build failures seemed to be caused by an unrelated HTTP error

**Review by jezdez — APPROVED** [review]

**Pull request merged**